File: Projetos/ACBrLib/Demos/C#/PosPrinter/Import/Dinamico/ACBrPosPrinter.cs
using System;
using System.IO;
using System.Runtime.InteropServices;
using System.Text;
using ACBrLib.Core;
using ACBrLib.Core.PosPrinter;

namespace ACBrLibPosPrinter
{
    /// <inheritdoc />
    public sealed class ACBrPosPrinter : ACBrLibHandle
    {
        #region InnerTypes

        private class Delegates
        {
            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_Inicializar(string eArqConfig, string eChaveCrypt);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_Finalizar();

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_Nome(StringBuilder buffer, ref int bufferSize);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_Versao(StringBuilder buffer, ref int bufferSize);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_UltimoRetorno(StringBuilder buffer, ref int bufferSize);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_ConfigLer(string eArqConfig);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_ConfigGravar(string eArqConfig);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_ConfigLerValor(string eSessao, string eChave, StringBuilder buffer, ref int bufferSize);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_ConfigGravarValor(string eSessao, string eChave, string valor);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_Ativar();

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_Desativar();

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_Imprimir(string aString, bool pulaLinha, bool decodificarTags, bool codificarPagina, int copias);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_ImprimirLinha(string aString);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_ImprimirCmd(string aString);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_ImprimirTags();

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_TxRx(string aString, byte bytesToRead, int aTimeOut, bool waitForTerminator, StringBuilder buffer, ref int bufferSize);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_Zerar();

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_InicializarPos();

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_Reset();

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_PularLinhas(int numLinhas);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_CortarPapel(bool parcial);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_AbrirGaveta();

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_LerInfoImpressora(StringBuilder buffer, ref int bufferSize);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_LerStatusImpressora(int tentativas, ref int status);

            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
            public delegate int POS_RetornarTags(StringBuilder buffer, ref int bufferSize, bool incluiAjuda);
        }

        #endregion InnerTypes

        #region Fields

        private const int BUFFER_LEN = 256;

        #endregion Fields

        #region Constructors

        public ACBrPosPrinter(string eArqConfig = "", string eChaveCrypt = "") : base(Environment.Is64BitProcess ? "ACBrPosPrinter64.dll" : "ACBrPosPrinter32.dll")
        {
            var inicializar = GetMethod<Delegates.POS_Inicializar>();
            var ret = ExecuteMethod(() => inicializar(ToUTF8(eArqConfig), ToUTF8(eChaveCrypt)));

            CheckResult(ret);
        }

        #endregion Constructors

        #region Properties

        public string Nome
        {
            get
            {
                var bufferLen = BUFFER_LEN;
                var buffer = new StringBuilder(bufferLen);

                var method = GetMethod<Delegates.POS_Nome>();
                var ret = ExecuteMethod(() => method(buffer, ref bufferLen));

                CheckResult(ret);

                return ProcessResult(buffer, bufferLen);
            }
        }

        public string Versao
        {
            get
            {
                var bufferLen = BUFFER_LEN;
                var buffer = new StringBuilder(bufferLen);

                var method = GetMethod<Delegates.POS_Versao>();
                var ret = ExecuteMethod(() => method(buffer, ref bufferLen));

                CheckResult(ret);

                return ProcessResult(buffer, bufferLen);
            }
        }

        #endregion Properties

        #region Configurações

        #region Ini

        public void ConfigGravar(string eArqConfig = "ACBrLib.ini")
        {
            var gravarIni = GetMethod<Delegates.POS_ConfigGravar>();
            var ret = ExecuteMethod(() => gravarIni(ToUTF8(eArqConfig)));

            CheckResult(ret);
        }

        public void ConfigLer(string eArqConfig = "ACBrLib.ini")
        {
            var lerIni = GetMethod<Delegates.POS_ConfigLer>();
            var ret = ExecuteMethod(() => lerIni(ToUTF8(eArqConfig)));

            CheckResult(ret);
        }

        public T ConfigLerValor<T>(ACBrSessao eSessao, string eChave)
        {
            var method = GetMethod<Delegates.POS_ConfigLerValor>();

            var bufferLen = BUFFER_LEN;
            var pValue = new StringBuilder(bufferLen);
            var ret = ExecuteMethod(() => method(ToUTF8(eSessao.ToString()), ToUTF8(eChave), pValue, ref bufferLen));
            CheckResult(ret);

            var value = ProcessResult(pValue, bufferLen);
            return ConvertValue<T>(value);
        }

        public void ConfigGravarValor(ACBrSessao eSessao, string eChave, object value)
        {
            if (value == null) return;

            var method = GetMethod<Delegates.POS_ConfigGravarValor>();
            var propValue = ConvertValue(value);

            var ret = ExecuteMethod(() => method(ToUTF8(eSessao.ToString()), ToUTF8(eChave), ToUTF8(propValue)));
            CheckResult(ret);
        }

        #endregion Ini

        #region Ativar

        public void Ativar()
        {
            var method = GetMethod<Delegates.POS_Ativar>();
            var ret = ExecuteMethod(() => method());

            CheckResult(ret);
        }

        public void Desativar()
        {
            var method = GetMethod<Delegates.POS_Desativar>();
            var ret = ExecuteMethod(() => method());

            CheckResult(ret);
        }

        #endregion Ativar

        #region Diversos

        public void Zerar()
        {
            var method = GetMethod<Delegates.POS_Zerar>();
            var ret = ExecuteMethod(() => method());

            CheckResult(ret);
        }

        public void Inicializar()
        {
            var method = GetMethod<Delegates.POS_InicializarPos>();
            var ret = ExecuteMethod(() => method());

            CheckResult(ret);
        }

        public void Reset()
        {
            var method = GetMethod<Delegates.POS_Reset>();
            var ret = ExecuteMethod(() => method());

            CheckResult(ret);
        }

        public void PularLinhas(int numLinhas = 0)
        {
            var method = GetMethod<Delegates.POS_PularLinhas>();
            var ret = ExecuteMethod(() => method(numLinhas));

            CheckResult(ret);
        }

        public void CortarPapel(bool parcial = false)
        {
            var method = GetMethod<Delegates.POS_CortarPapel>();
            var ret = ExecuteMethod(() => method(parcial));

            CheckResult(ret);
        }

        public void AbrirGaveta()
        {
            var method = GetMethod<Delegates.POS_AbrirGaveta>();
            var ret = ExecuteMethod(() => method());

            CheckResult(ret);
        }

        public string LerInfoImpressora()
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Delegates.POS_LerInfoImpressora>();
            var ret = ExecuteMethod(() => method(buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        public ACBrPosTipoStatus LerStatusImpressora(int tentativas = 1)
        {
            var status = 0;
            var method = GetMethod<Delegates.POS_LerStatusImpressora>();
            var ret = ExecuteMethod(() => method(tentativas, ref status));

            CheckResult(ret);

            return (ACBrPosTipoStatus)status;
        }

        public string[] RetornarTags(bool incluiAjuda = true)
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Delegates.POS_RetornarTags>();
            var ret = ExecuteMethod(() => method(buffer, ref bufferLen, incluiAjuda));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen).Split('|');
        }

        #endregion Diversos

        #region Imprimir

        public void Imprimir(string aString = "", bool pulaLinha = false, bool decodificarTags = true, bool codificarPagina = true, int copias = 1)
        {
            var method = GetMethod<Delegates.POS_Imprimir>();
            var ret = ExecuteMethod(() => method(ToUTF8(aString), pulaLinha, decodificarTags, codificarPagina, copias));

            CheckResult(ret);
        }

        public void ImprimirLinha(string aString)
        {
            var method = GetMethod<Delegates.POS_ImprimirLinha>();
            var ret = ExecuteMethod(() => method(ToUTF8(aString)));

            CheckResult(ret);
        }

        public void ImprimirCmd(string aString)
        {
            var method = GetMethod<Delegates.POS_ImprimirCmd>();
            var ret = ExecuteMethod(() => method(ToUTF8(aString)));

            CheckResult(ret);
        }

        public void ImprimirTags()
        {
            var method = GetMethod<Delegates.POS_ImprimirTags>();
            var ret = ExecuteMethod(() => method());

            CheckResult(ret);
        }

        public string TxRx(string aString, byte bytesToRead = 1, int aTimeOut = 500, bool waitForTerminator = false)
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Delegates.POS_TxRx>();
            var ret = ExecuteMethod(() => method(ToUTF8(aString), bytesToRead, aTimeOut, waitForTerminator, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        #endregion Imprimir

        #region Private Methods

        protected override void FinalizeLib()
        {
            var finalizar = GetMethod<Delegates.POS_Finalizar>();
            var codRet = ExecuteMethod(() => finalizar());
            CheckResult(codRet);
        }

        protected override void InitializeMethods()
        {
            AddMethod<Delegates.POS_Inicializar>("POS_Inicializar");
            AddMethod<Delegates.POS_Finalizar>("POS_Finalizar");
            AddMethod<Delegates.POS_Nome>("POS_Nome");
            AddMethod<Delegates.POS_Versao>("POS_Versao");
            AddMethod<Delegates.POS_UltimoRetorno>("POS_UltimoRetorno");
            AddMethod<Delegates.POS_ConfigLer>("POS_ConfigLer");
            AddMethod<Delegates.POS_ConfigGravar>("POS_ConfigGravar");
            AddMethod<Delegates.POS_ConfigLerValor>("POS_ConfigLerValor");
            AddMethod<Delegates.POS_ConfigGravarValor>("POS_ConfigGravarValor");
            AddMethod<Delegates.POS_Ativar>("POS_Ativar");
            AddMethod<Delegates.POS_Desativar>("POS_Desativar");
            AddMethod<Delegates.POS_Imprimir>("POS_Imprimir");
            AddMethod<Delegates.POS_ImprimirLinha>("POS_ImprimirLinha");
            AddMethod<Delegates.POS_ImprimirCmd>("POS_ImprimirCmd");
            AddMethod<Delegates.POS_ImprimirTags>("POS_ImprimirTags");
            AddMethod<Delegates.POS_TxRx>("POS_TxRx");
            AddMethod<Delegates.POS_Zerar>("POS_Zerar");
            AddMethod<Delegates.POS_InicializarPos>("POS_InicializarPos");
            AddMethod<Delegates.POS_Reset>("POS_Reset");
            AddMethod<Delegates.POS_PularLinhas>("POS_PularLinhas");
            AddMethod<Delegates.POS_CortarPapel>("POS_CortarPapel");
            AddMethod<Delegates.POS_AbrirGaveta>("POS_AbrirGaveta");
            AddMethod<Delegates.POS_LerInfoImpressora>("POS_LerInfoImpressora");
            AddMethod<Delegates.POS_LerStatusImpressora>("POS_LerStatusImpressora");
            AddMethod<Delegates.POS_RetornarTags>("POS_RetornarTags");
        }

        protected override string GetUltimoRetorno(int iniBufferLen = 0)
        {
            var bufferLen = iniBufferLen < 1 ? BUFFER_LEN : iniBufferLen;
            var buffer = new StringBuilder(bufferLen);
            var ultimoRetorno = GetMethod<Delegates.POS_UltimoRetorno>();

            if (iniBufferLen < 1)
            {
                ExecuteMethod(() => ultimoRetorno(buffer, ref bufferLen));
                if (bufferLen <= BUFFER_LEN) return FromUTF8(buffer);

                buffer.Capacity = bufferLen;
            }

            ExecuteMethod(() => ultimoRetorno(buffer, ref bufferLen));
            return FromUTF8(buffer);
        }

        #endregion Private Methods

        #endregion Configurações
    }
}

File: Projetos/ACBrLib/Demos/Java/PosPrinter/Imports/ACBrLibPosPrinter/src/com/acbr/posprinter/ACBrPosPrinter.java
  package com.acbr.posprinter;

import com.acbr.ACBrSessao;
import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.Platform;
import com.sun.jna.ptr.IntByReference;
import java.io.File;

import java.nio.ByteBuffer;
import java.nio.charset.Charset;
import java.nio.file.Paths;
import java.util.Set;

public final class ACBrPosPrinter {

    public interface PosPrinterLib extends Library {

        public static String JNA_LIBRARY_NAME = LibraryLoader.getLibraryName();
        public static PosPrinterLib INSTANCE = LibraryLoader.getInstance();

        class LibraryLoader {

            private static String library = "";
            private static PosPrinterLib instance = null;

            public static String getLibraryName() {
                if (library.isEmpty()) {
                    library = Platform.is64Bit() ? "ACBrPosPrinter64" : "ACBrPosPrinter32";
                }
                return library;
            }

            public static PosPrinterLib getInstance() {
                if (instance == null) {
                    instance = (PosPrinterLib) Native.synchronizedLibrary((Library) Native.loadLibrary(JNA_LIBRARY_NAME, PosPrinterLib.class));
                }
                return instance;
            }
        }

        int POS_Inicializar(String eArqConfig, String eChaveCrypt);

        int POS_Finalizar();

        int POS_Nome(ByteBuffer buffer, IntByReference bufferSize);

        int POS_Versao(ByteBuffer buffer, IntByReference bufferSize);

        int POS_UltimoRetorno(ByteBuffer buffer, IntByReference bufferSize);

        int POS_ConfigLer(String eArqConfig);

        int POS_ConfigGravar(String eArqConfig);

        int POS_ConfigLerValor(String eSessao, String eChave, ByteBuffer buffer, IntByReference bufferSize);

        int POS_ConfigGravarValor(String eSessao, String eChave, String valor);

        int POS_Ativar();

        int POS_Desativar();

        int POS_Imprimir(String aString, boolean pulaLinha, boolean decodificarTags, boolean codificarPagina, int copias);

        int POS_ImprimirLinha(String aString);

        int POS_ImprimirCmd(String aString);

        int POS_ImprimirTags();
        
        int POS_ImprimirImagemArquivo(String APath);
        
        int POS_GravarLogoArquivo(String APath, int nAKC1, int nAKC2);
        
        int POS_ImprimirLogo(int nAKC1, int nAKC2, int nFatorX, int nFatorY);
        
        int POS_ApagarLogo(int nAKC1, int nAKC2);

        int POS_TxRx(String aString, byte bytesToRead, int aTimeOut, boolean waitForTerminator, ByteBuffer buffer, IntByReference bufferSize);

        int POS_Zerar();

        int POS_InicializarPos();

        int POS_Reset();

        int POS_PularLinhas(int numLinhas);

        int POS_CortarPapel(boolean parcial);

        int POS_AbrirGaveta();

        int POS_LerInfoImpressora(ByteBuffer buffer, IntByReference bufferSize);

        int POS_LerStatusImpressora(int tentativas, IntByReference status);

        int POS_RetornarTags(ByteBuffer buffer, IntByReference bufferSize, boolean incluiAjuda);

    }

    private final static int STR_BUFFER_LEN = 256;
    private final static Charset UTF8 = Charset.forName("UTF-8");

    public ACBrPosPrinter() throws Exception {
        File iniFile = Paths.get(System.getProperty("user.dir"), "ACBrLib.ini").toFile();
        if (!iniFile.exists()) {
            iniFile.createNewFile();
        }

        int ret = PosPrinterLib.INSTANCE.POS_Inicializar(toUTF8(iniFile.getAbsolutePath()), toUTF8(""));
        checkResult(ret);
    }

    public ACBrPosPrinter(String eArqConfig, String eChaveCrypt) throws Exception {
        int ret = PosPrinterLib.INSTANCE.POS_Inicializar(toUTF8(eArqConfig), toUTF8(eChaveCrypt));
        checkResult(ret);
    }

    @Override
    protected void finalize() throws Throwable {
        try {
            int ret = PosPrinterLib.INSTANCE.POS_Finalizar();
            checkResult(ret);
        } finally {
            super.finalize();
        }
    }

    public String nome() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = PosPrinterLib.INSTANCE.POS_Nome(buffer, bufferLen);
        checkResult(ret);

        return fromUTF8(buffer, bufferLen.getValue());
    }

    public String versao() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = PosPrinterLib.INSTANCE.POS_Versao(buffer, bufferLen);
        checkResult(ret);

        return fromUTF8(buffer, bufferLen.getValue());
    }

    public void configLer() throws Exception {
        configLer("");
    }

    public void configLer(String eArqConfig) throws Exception {
        int ret = PosPrinterLib.INSTANCE.POS_ConfigLer(toUTF8(eArqConfig));
        checkResult(ret);
    }

    public void configGravar() throws Exception {
        configGravar("");
    }

    public void configGravar(String eArqConfig) throws Exception {
        int ret = PosPrinterLib.INSTANCE.POS_ConfigGravar(toUTF8(eArqConfig));
        checkResult(ret);
    }

    public String configLerValor(ACBrSessao eSessao, String eChave) throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = PosPrinterLib.INSTANCE.POS_ConfigLerValor(toUTF8(eSessao.name()), toUTF8(eChave), buffer, bufferLen);
        checkResult(ret);

        return processResult(buffer, bufferLen);
    }

    public void configGravarValor(ACBrSessao eSessao, String eChave, Object value) throws Exception {
        int ret = PosPrinterLib.INSTANCE.POS_ConfigGravarValor(toUTF8(eSessao.name()), toUTF8(eChave), toUTF8(value.toString()));
        checkResult(ret);
    }

    public void ativar() throws Exception {
        int ret = PosPrinterLib.INSTANCE.POS_Ativar();
        checkResult(ret);
    }

    public void desativar() throws Exception {
        int ret = PosPrinterLib.INSTANCE.POS_Desativar();
        checkResult(ret);
    }

    public void zerar() throws Exception {
        int ret = PosPrinterLib.INSTANCE.POS_Zerar();
        checkResult(ret);
    }

    public void inicializar() throws Exception {
        int ret = PosPrinterLib.INSTANCE.POS_InicializarPos();
        checkResult(ret);
    }

    public void reset() throws Exception {
        int ret = PosPrinterLib.INSTANCE.POS_Reset();
        checkResult(ret);
    }

    public void pularLinhas() throws Exception {
        pularLinhas(0);
    }

    public void pularLinhas(int numLinhas) throws Exception {
        int ret = PosPrinterLib.INSTANCE.POS_PularLinhas(numLinhas);
        checkResult(ret);
    }

    public void cortarPapel() throws Exception {
        cortarPapel(false);
    }

    public void cortarPapel(boolean parcial) throws Exception {
        int ret = PosPrinterLib.INSTANCE.POS_CortarPapel(parcial);
        checkResult(ret);
    }

    public void abrirGaveta() throws Exception {
        int ret = PosPrinterLib.INSTANCE.POS_AbrirGaveta();
        checkResult(ret);
    }

    public String lerInfoImpressora() throws Exception {

        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = PosPrinterLib.INSTANCE.POS_LerInfoImpressora(buffer, bufferLen);
        checkResult(ret);

        return processResult(buffer, bufferLen);
    }

    public Set<ACBrPosTipoStatus> lerStatusImpressora() throws Exception {
        return lerStatusImpressora(1);
    }

    public Set<ACBrPosTipoStatus> lerStatusImpressora(int tentativas) throws Exception {
        IntByReference status = new IntByReference(0);
        int ret = PosPrinterLib.INSTANCE.POS_LerStatusImpressora(tentativas, status);
        checkResult(ret);

        return ACBrPosTipoStatus.valueOf(status.getValue());
    }

    public String[] retornarTags() throws Exception {
        return retornarTags(true);
    }

    public String[] retornarTags(boolean incluiAjuda) throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = PosPrinterLib.INSTANCE.POS_RetornarTags(buffer, bufferLen, incluiAjuda);
        checkResult(ret);

        return processResult(buffer, bufferLen).split("|");
    }

    public void imprimir(String aString) throws Exception {
        imprimir(aString, false, true, true, 1);
    }

    public void imprimir(String aString, boolean pulaLinha, boolean decodificarTags, boolean codificarPagina, int copias) throws Exception {
        int ret = PosPrinterLib.INSTANCE.POS_Imprimir(toUTF8(aString), pulaLinha, decodificarTags, codificarPagina, copias);
        checkResult(ret);
    }

    public void imprimirLinha(String aString) throws Exception {
        int ret = PosPrinterLib.INSTANCE.POS_ImprimirLinha(toUTF8(aString));
        checkResult(ret);
    }

    public void imprimirCmd(String aString) throws Exception {
        int ret = PosPrinterLib.INSTANCE.POS_ImprimirCmd(toUTF8(aString));
        checkResult(ret);
    }

    public void imprimirTags() throws Exception {
        int ret = PosPrinterLib.INSTANCE.POS_ImprimirTags();
        checkResult(ret);
    }
    
    public void imprimirImagemArquivo(String aPath) throws Exception {
        int ret = PosPrinterLib.INSTANCE.POS_ImprimirImagemArquivo(toUTF8(aPath));
        checkResult(ret);
    }
    
    public void gravarLogoArquivo(String aPath) throws Exception {
        gravarLogoArquivo(aPath, -1, -1);
    }
    
    public void gravarLogoArquivo(String aPath, int nAKC1, int nAKC2) throws Exception {
        int ret = PosPrinterLib.INSTANCE.POS_GravarLogoArquivo(toUTF8(aPath), nAKC1, nAKC2);
        checkResult(ret);
    }
    
    public void imprimirLogo() throws Exception {
        imprimirLogo(-1, -1, -1, -1);
    }
    
    public void imprimirLogo(int nAKC1, int nAKC2, int nFatorX, int nFatorY) throws Exception {
        int ret = PosPrinterLib.INSTANCE.POS_ImprimirLogo(nAKC1, nAKC2, nFatorX, nFatorY);
        checkResult(ret);
    }
    
    public void apagarLogo() throws Exception {
        apagarLogo(-1, -1);
    }
    
    public void apagarLogo(int nAKC1, int nAKC2) throws Exception {
        int ret = PosPrinterLib.INSTANCE.POS_ApagarLogo(nAKC1, nAKC2);
        checkResult(ret);
    }

    public String txRx(String aString) throws Exception {
        return txRx(aString, (byte) 1, 500, false);
    }

    public String txRx(String aString, byte bytesToRead, int aTimeOut, boolean waitForTerminator) throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = PosPrinterLib.INSTANCE.POS_TxRx(toUTF8(aString), bytesToRead, aTimeOut, waitForTerminator, buffer, bufferLen);
        checkResult(ret);

        return processResult(buffer, bufferLen);
    }

    private static String toUTF8(String value) {
        return new String(value.getBytes(UTF8));
    }

    private static String fromUTF8(ByteBuffer buffer, int len) {
        return new String(buffer.array(), 0, len, UTF8);
    }

    private static void checkResult(int result) throws Exception {
        if (result == 0) {
            return;
        }

        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        PosPrinterLib.INSTANCE.POS_UltimoRetorno(buffer, bufferLen);
        throw new Exception(processResult(buffer, bufferLen));
    }

    private static String processResult(ByteBuffer buffer, IntByReference bufferLen) {
        if (bufferLen.getValue() <= STR_BUFFER_LEN) {
            return fromUTF8(buffer, bufferLen.getValue());
        }

        if (bufferLen.getValue() > STR_BUFFER_LEN) {
            buffer = ByteBuffer.allocate(bufferLen.getValue());
            PosPrinterLib.INSTANCE.POS_UltimoRetorno(buffer, bufferLen);
        }

        return fromUTF8(buffer, bufferLen.getValue());
    }
}
